### PR TITLE
Machine learning

### PR DESCRIPTION
**Pull request by MHeymann**
Create a dynamic module that does basic machine learning on the work done in the current singular session, and suggest 5 singular help pages to read from.

### CODE REVIEW

File: Singular/iparith.cc [hunk 1]
```diff
@@ -7993,7 +7994,9 @@ static BOOLEAN jjRESERVEDLIST0(leftv res, leftv)
 			k++;
 		}
 	}
-	// free the struct (not the list itself)
+	// free the struct (not the list entries itself, which were allocated
+	// by strdup)
+	omfree(bb_list->list);
```
Comment (MHeymann) [inline Singular/iparith.cc]: I realized that this was a memory leak...

**Comment by hannes14** [comment]
Singular/dyn_modules/machinelearning/machinelearning.cc requires an include <stdio.h>
before the include <readline/history.h>.

**Comment by hannes14** [comment]
>  predictHelp(); results in (after numpy installation):
Traceback (most recent call last):
  File "...lookuptable.py", line 16, in <module>
    from common.keyword_vector import get_vectors, read_dictionary
  File "....share/ml_python/common/keyword_vector.py", line 9, in <module>
    from sklearn.feature_extraction.text import CountVectorizer
ImportError: No module named sklearn.feature_extraction.text
Failed to load "common.lookuptable"
Initialise of ml failed.

**Comment by MHeymann** [comment]
Hi Hans!

I think this can be solved by installing sklearn. On debian Machines I
would do this with
$ apt install python-sklearn

Hans Schoenemann <notifications@github.com> schrieb am Di., 1. Okt. 2019,
16:08:

**Comment by MHeymann** [comment]
> Singular/dyn_modules/machinelearning/machinelearning.cc requires an include <stdio.h>
> before the include <readline/history.h>.

Fixed

**Comment by hannes14** [comment]
now (after apt install python-sklearn on a debian machine):
> predictHelp();
ImportError: No module named common.lookuptable
Failed to load "common.lookuptable"
Initialise of ml failed.

**Comment by MHeymann** [comment]
Hi Hans!

So I tried to think what this could be.  I haven't been able to reproduce
the error on my side.

Are you running the binary that gets built in the source tree after "make
-j 2", or did you run "make install" and run the installed binary? In fact,
could you send me some detailed build steps you followed for me to
reproduce the error?

**Comment by hannes14** [comment]
I do:
export CFLAGS="-O2 -pipe -fno-common -fno-exceptions"
export CXXFLAGS="-O2 -pipe -fno-common -fno-exceptions -fno-rtti"
./configure -C --enable-gfanlib --disable-static --with-python --enable-python_module --prefix=/tmp --with-flint=/usr/local
make install
and run /tmp/bin/Singular

**Pull request merged**